src/clp_ffi_py/ir/native/PySerializer.cpp

```diff
@@ -0,0 +1,576 @@
+        );
+        return nullptr;
+    }
+
+    auto* py_bytes_msgpack_map{py_reinterpret_cast<PyBytesObject>(msgpack_map)};
+    // Since the type is already checked, we can use the macro to avoid duplicated type checking.
+    auto const num_byte_serialized{self->serialize_log_event_from_msgpack_map(
+            {PyBytes_AS_STRING(py_bytes_msgpack_map),
+             static_cast<size_t>(PyBytes_GET_SIZE(py_bytes_msgpack_map))}
+    )};
+    if (false == num_byte_serialized.has_value()) {
+        return nullptr;
+    }
+
+    return PyLong_FromSsize_t(num_byte_serialized.value());
+}
+
+CLP_FFI_PY_METHOD auto PySerializer_get_num_bytes_serialized(PySerializer* self) -> PyObject* {
+    return PyLong_FromSsize_t(self->get_num_bytes_serialized());
+}
+
+CLP_FFI_PY_METHOD auto PySerializer_flush(PySerializer* self) -> PyObject* {
+    if (false == self->flush()) {
+        return nullptr;
+    }
+    Py_RETURN_NONE;
+}
+
+CLP_FFI_PY_METHOD auto PySerializer_close(PySerializer* self) -> PyObject* {
+    if (false == self->close()) {
+        return nullptr;
+    }
+    Py_RETURN_NONE;
+}
+
+CLP_FFI_PY_METHOD auto PySerializer_enter(PySerializer* self) -> PyObject* {
+    Py_INCREF(self);
+    return py_reinterpret_cast<PyObject>(self);
+}
+
+CLP_FFI_PY_METHOD auto
+PySerializer_exit(PySerializer* self, PyObject* args, PyObject* keywords) -> PyObject* {
+    static char keyword_exc_type[]{"exc_type"};
+    static char keyword_exc_value[]{"exc_value"};
+    static char keyword_traceback[]{"traceback"};
+    static char* keyword_table[]{
+            static_cast<char*>(keyword_exc_type),
+            static_cast<char*>(keyword_exc_value),
+            static_cast<char*>(keyword_traceback),
+            nullptr
+    };
+
+    PyObject* py_exc_type{};
+    PyObject* py_exc_value{};
+    PyObject* py_traceback{};
+    if (false
+        == static_cast<bool>(PyArg_ParseTupleAndKeywords(
+                args,
+                keywords,
+                "|OOO",
+                static_cast<char**>(keyword_table),
+                &py_exc_type,
+                &py_exc_value,
+                &py_traceback
+        )))
+    {
+        return nullptr;
+    }
+
+    // We don't do anything with the given exception. It is the caller's responsibility to raise
+    // the exceptions: https://docs.python.org/3/reference/datamodel.html#object.__exit__
+    if (false == self->close()) {
+        return nullptr;
+    }
+
+    Py_RETURN_NONE;
+}
+
+CLP_FFI_PY_METHOD auto PySerializer_dealloc(PySerializer* self) -> void {
+    PyErrGuard const err_guard;
+
+    if (false == self->is_closed()) {
+        if (0
+            != PyErr_WarnEx(
+                    PyExc_ResourceWarning,
+                    "`Serializer.close()` is not called before object destruction. Forget to call "
+                    "`close` will leave the stream incomplete, and potentially resulting in data "
+                    "loss due to data buffering",
```

Comment:
Nit:
and I believe we should keep the space at the beginning of the line
```suggestion
                    "`Serializer.close()` is not called before object destruction, which will leave"
                    " the stream incomplete, and potentially resulting in data"
                    " loss due to data buffering",
```

Reply:
Yeah make sense, mb